file_writer_control/CommandChannel.py

```diff
@@ -0,0 +1,217 @@
+            )  # Roughly 300MB
+            break
+        except NoBrokersAvailable:
+            pass
+        if not in_queue.empty():
+            new_msg = in_queue.get()
+            if new_msg == "exit":
+                return
+    while True:
+        for message in consumer:
+            status_tracker.process_message(message.value)
+        status_tracker.check_for_lost_connections()
+        if not in_queue.empty():
+            new_msg = in_queue.get()
+            if new_msg == "exit":
+                break
+    consumer.close(True)
+
+
+class CommandChannel(object):
+    """
+    A class that implements the functionality for receiving and interpreting messages that are published to the
+    Kafka command topic of a pool of file-writers.
+    """
+
+    def __init__(self, command_topic_url: str):
+        """
+        Constructor.
+        :param command_topic_url: The url of the Kafka topic to where the file-writer status/command messages are published.
+        """
+        kafka_address = KafkaTopicUrl(command_topic_url)
+        self.status_queue = Queue()
+        self.to_thread_queue = Queue()
+        thread_kwargs = {
+            "host_port": kafka_address.host_port,
+            "topic": kafka_address.topic,
+            "in_queue": self.to_thread_queue,
+            "out_queue": self.status_queue,
+        }
+        self.map_of_workers: Dict[str, WorkerStatus] = {}
+        self.map_of_jobs: Dict[str, JobStatus] = {}
+        self.map_of_commands: Dict[str, CommandStatus] = {}
+        self.run_thread = True
+        self.thread = threading.Thread(
+            target=thread_function, daemon=True, kwargs=thread_kwargs
+        )
+        self.thread.start()
+
+        def do_exit():
+            self.stop_thread()
+
+        atexit.register(do_exit)
+
+    def add_job_id(self, job_id: str):
+        """
+        Add a job identifier to the list of known jobs before it has been encountered on the command topic.
+        :param job_id: The identifier of the new job.
+        """
+        if job_id not in self.map_of_jobs:
+            self.map_of_jobs[job_id] = JobStatus(job_id)
+
+    def add_command_id(self, job_id: str, command_id: str):
+        """
+        Add a command identifier to the list of known commands before it has been encountered on the command topic.
+        :param job_id: The job identifier of the new command.
+        :param command_id: The identifier of the new command.
+        """
+        if command_id not in self.map_of_commands:
+            self.map_of_commands[command_id] = CommandStatus(job_id, command_id)
+
+    def stop_thread(self):
+        """
+        Stop the thread that is continuously getting command topic messages in the background. Should only be called if
+        we are about to get rid of the current instance of CommandChannel.
+        """
+        self.to_thread_queue.put("exit")
+        try:
+            self.thread.join()
+        except RuntimeError:
+            pass
```

Comment:
See the comment above.

Reply:
If the thread has not yet been started when this function is called, it is bad form (in my opinion) to throw an exception. I have created a comment that explains why we `pass` here.